keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1194 @@
+	// The labels of the evacuation object will be copied from pod's .metadata.labels.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec defines the evacuation.
+	// https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +optional
+	Spec EvacuationSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status represents the most recently observed status of the evacuation.
+	// Populated by the current evacuator.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +optional
+	Status EvacuationStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// EvacuationSpec is a specification of an Evacuation.
+type EvacuationSpec struct {
+	// PodRef references a pod that is subject to evacuation.
+	// This field is required and immutable.
+	// +optional
+	PodRef *PodReference `json:"podRef,omitempty" protobuf:"bytes,1,opt,name=podRef"`
+
+	// AcceptDeadlineSeconds is a maximum amount of time an evacuator should take to respond to an
+	// evacuation. It should respond by setting .status.accepted to true or false, and providing
+	// .status.evacuatorRef and .status.evacuationProgressTimestamp. The evacuator should then
+	// periodically update .status.evacuationProgressTimestamp.
+	// If the .status.accepted is not updated within the duration of .spec.acceptDeadlineSeconds,
+	// the evacuated pod will be evicted using the Eviction API.
+	//
+	// The minimum value is 60 (1m) and the maximum value is 3600 (1h).
+	// The value must be less than or equal to .spec.progressDeadlineSeconds.
+	// This field is required and immutable.
+	AcceptDeadlineSeconds int32 `json:"acceptDeadlineSeconds" protobuf:"varint,2,opt,name=acceptDeadlineSeconds"`
+	
+	// ProgressDeadlineSeconds is a maximum amount of time an evacuator should take to report on a
+	// progress by updating the .status.evacuationProgressTimestamp.
+	// If the .status.evacuationProgressTimestamp is not updated within the duration of
+	// .spec.progressDeadlineSeconds, the evacuated pod will be evicted using the Eviction API.
+	//
+	// The minimum value is 60 (1m) and the maximum value is 43200 (12h).
+	// The value must be greater than or equal to .spec.acceptDeadlineSeconds.
+	// This field is required and immutable.
+	ProgressDeadlineSeconds int32 `json:"progressDeadlineSeconds" protobuf:"varint,3,opt,name=progressDeadlineSeconds"`
+}
+
+// PodReference contains enough information to locate the referenced pod.
+type PodReference struct {
+	// Name of the pod.
+	// This field is required.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// Namespace of the pod.
+	// This field is required.
+	Namespace string `json:"namespace" protobuf:"bytes,2,opt,name=namespace"`
```

Comment:
Does this mean that Evacuation is not namespaced? If so, how will we restrict access to Evacuations and prevent cross-namespace abuse?

Reply:
Yeah we do not have to mention the namespace for the pod as the Evacuation should be namespaced.

It might be possible to introduce the evacuation concept even for other objects in the future. I am not sure if that is desirable, so I am targeting only pods at this time.

Reply:
added to non-goals